packages/cli/src/Server.ts

```diff
@@ -1801,42 +1766,41 @@ class App {
 				scriptsString += phLoadingScript;
 			}
 
-			const firstLinkedScriptSegment = '<link href="/js/';
-			readIndexFile = readIndexFile.replace(
-				firstLinkedScriptSegment,
-				scriptsString + firstLinkedScriptSegment,
-			);
-
-			// Serve the altered index.html file separately
-			this.app.get(`/index.html`, async (req: express.Request, res: express.Response) => {
-				res.send(readIndexFile);
-			});
+			const editorUiDistDir = pathJoin(pathDirname(require.resolve('n8n-editor-ui')), 'dist');
+			const generatedStaticDir = pathJoin(__dirname, '../public');
 
-			this.app.get('/assets/*.css', async (req: express.Request, res: express.Response) => {
-				res.type('text/css').send(cssFiles[req.url]);
-			});
-
-			this.app.get('/assets/*.js', async (req: express.Request, res: express.Response) => {
-				res.type('text/javascript').send(jsFiles[req.url]);
-			});
-
-			// Serve the website
+			const firstLinkedScriptSegment = '<link href="/js/';
 			this.app.use(
 				'/',
-				express.static(pathJoin(pathDirname(editorUiPath), 'dist'), {
-					index: 'index.html',
-					setHeaders: (res, path) => {
-						if (res.req && res.req.url === '/index.html') {
-							// Set last modified date manually to n8n start time so
-							// that it hopefully refreshes the page when a new version
-							// got used
-							res.setHeader('Last-Modified', startTime);
+				express.static(generatedStaticDir),
+				async (req, res, next) => {
+					const fileName = req.url.slice(0);
+					const filePath = pathJoin(editorUiDistDir, fileName);
+					if (/(index.html)|.*\.(js|css)/.test(filePath) && existsSync(filePath)) {
+						const srcFile = await readFile(filePath, 'utf8');
+						let payload = srcFile.replace(basePathRegEx, n8nPath);
+						if (req.url === '/index.html') {
+							payload = payload
+								.replace(/\/favicon.ico/g, `${n8nPath}favicon.ico`)
+								.replace(firstLinkedScriptSegment, scriptsString + firstLinkedScriptSegment);
 						}
-					},
-				}),
+						const destFile = pathJoin(generatedStaticDir, fileName);
+						await mkdir(pathDirname(destFile), { recursive: true });
+						await writeFile(destFile, payload, 'utf-8');
+						send(req, destFile).pipe(res);
+						return;
+					}
```

Comment:
Would it make sense to generate the pre-processed `public` folder beforehand and just serve assets from the pre-processed dir?

This way we could eliminate these checks on every asset request.

Reply:
good catch. I'll try to give it a shot.

Reply:
Done :heavy_check_mark: 
I also discovered today that we are not returning `Cache-Control` headers on these static assets on master ([Ticket](https://linear.app/n8n/issue/N8N-4805)). Switching to pre-generating these and using `express.static` fixed that as well.